File: Uva/1174.py
def eachSet(graph):
    input()
    m = int(input())
    n = int(input())
    graph[:] = []
    dictonary = {}
    point = 1
    # print(graph)
    for _ in range(n):
        x, y, w = input().split(" ")
        w = int(w)
        if(dictonary.get(x) == None):
            dictonary[x] = point
            x = point
            point += 1
        else:
            x = dictonary.get(x)
        if(dictonary.get(y) == None):
            dictonary[y] = point
            y = point
            point += 1
        else:
            y = dictonary.get(y)
        graph.append((x, y, w))


def check(mark, r):
    if(mark[r] == r):
        return r
    return check(mark, mark[r])


def kruskal(graph):
    graph.sort(key=lambda x: x[2:])
    mark = list(range(0, len(graph)+1))
    # print(mark)

    total = 0
    for it in graph:
        x = check(mark, it[0])
        y = check(mark, it[1])
        if x != y:
            mark[x] = y
            total += it[2]
    print(total)


def solve():
    graph = []
    test = int(input())
    for t in range(test):
        eachSet(graph)
        if t:
            print()
        kruskal(graph)


solve()

File: Uva/1174.cpp
#include <bits/stdc++.h>
using namespace std;
#define Size 2005
struct node
{
    int x, y, w;
    bool operator<(const node &p) const
    {
        return w < p.w;
    }
} temp;
int mark[Size];
vector<node> graph;
int Find(int r)
{
    if (mark[r] == r)
        return r;
    return mark[r] = Find(mark[r]);
}
void karuskal(int n)
{
    int i, sum = 0, x, y;
    sort(graph.begin(), graph.end());
    for (i = 0; i <= n; i++)
        mark[i] = i;

    for (auto it : graph)
    {
        x = Find(it.x);
        y = Find(it.y);
        if (x != y)
        {
            mark[x] = y;
            sum += it.w;
        }
    }
    printf("%d\n", sum);
}
int main(){
    // freopen("input.txt", "r", stdin);
    int test, n, m;
    scanf("%d", &test);
    for(int t=0; t<test; t++){
        scanf("%d%d", &n, &m);
        graph.clear();
        map<string, int>temporarySet;
        int point=1;
        string a, b;

        for(int i=0; i<m; i++){
            cin>>a>>b;
            scanf("%d", &temp.w);
            if(temporarySet[a])
                temp.x = temporarySet[a];
            else
            {
                temporarySet[a] = point;
                temp.x = point;
                point++;
            }
            if(temporarySet[b])
                temp.y = temporarySet[b];
            else
            {
                temporarySet[b] = point;
                temp.y = point;
                point++;
            }
            graph.push_back(temp);
        }
        if(t)
            puts("");
        karuskal(n);
    }
}